File: ai_multicolony/api/routes/ecosystem.py
"""
Ecosystem Routes — Aggregated API endpoints for all ecosystem services.

Provides cross-service status checks, data proxying, and a unified
overview dashboard endpoint.
"""

from __future__ import annotations

from typing import Any, Dict

from fastapi import APIRouter

from ai_multicolony.config.logging_config import get_logger

logger = get_logger(__name__)


def create_router() -> APIRouter:
    """Create the ecosystem routes router."""
    router = APIRouter()

    @router.get("/status", summary="Overall ecosystem health")
    async def ecosystem_status() -> Dict[str, Any]:
        """Check health of all ecosystem services.

        Returns a status object for each integrated package:
        Crucix OSINT, HermesQuant Trading, Autonomous Organism.
        """
        result: Dict[str, Any] = {
            "ecosystem": "AI-MultiColony-Ecosystem",
            "version": "3.0.0",
            "services": {},
        }

        # Check Crucix OSINT
        try:
            from ai_multicolony.integrations.crucix_client import CrucixClient

            crucix = CrucixClient()
            healthy = await crucix.is_healthy()
            result["services"]["crucix"] = {
                "status": "healthy" if healthy else "unhealthy",
                "type": "osint_intelligence",
                "port": 3117,
            }
            await crucix.close()
        except Exception as exc:
            result["services"]["crucix"] = {
                "status": "unavailable",
                "error": str(exc),
            }

        # Check HermesQuant
        try:
            from ai_multicolony.integrations.hermes_bridge import HermesQuantBridge

            hermes = HermesQuantBridge()
            available = await hermes.is_available()
            result["services"]["hermes_quant"] = {
                "status": "available" if available else "unavailable",
                "type": "quantitative_trading",
            }
        except Exception as exc:
            result["services"]["hermes_quant"] = {
                "status": "unavailable",
                "error": str(exc),
            }

        # Check Autonomous Organism
        try:
            from ai_multicolony.integrations.organism_bridge import OrganismBridge
            import os

            organism = OrganismBridge(
                supabase_url=os.getenv("SUPABASE_URL", ""),
                supabase_anon_key=os.getenv("SUPABASE_ANON_KEY", ""),
            )
            available = await organism.is_available()
            result["services"]["autonomous_organism"] = {
                "status": "available" if available else "unavailable",
                "type": "autonomous_organism",
            }
            await organism.close()
        except Exception as exc:
            result["services"]["autonomous_organism"] = {
                "status": "unavailable",
                "error": str(exc),
            }

        # Overall status
        all_healthy = all(
            s.get("status") in ("healthy", "available")
            for s in result["services"].values()
        )
        result["overall"] = "healthy" if all_healthy else "degraded"

        return result

    @router.get("/crucix", summary="Crucix OSINT data")
    async def crucix_data() -> Dict[str, Any]:
        """Proxy to Crucix OSINT sweep data."""
        try:
            from ai_multicolony.integrations.crucix_client import CrucixClient

            client = CrucixClient()
            data = await client.get_sweep_data()
            health = await client.get_health()
            await client.close()
            return {"sweep": data.model_dump(), "health": health.model_dump()}
        except Exception as exc:
            logger.error("ecosystem_crucix_error", error=str(exc))
            return {"error": str(exc), "status": "unavailable"}

    @router.get("/hermes", summary="HermesQuant trading status")
    async def hermes_status() -> Dict[str, Any]:
        """HermesQuant trading engine status and portfolio data."""
        try:
            from ai_multicolony.integrations.hermes_bridge import HermesQuantBridge

            bridge = HermesQuantBridge()
            portfolio = await bridge.get_portfolio_status()
            available = await bridge.is_available()
            return {
                "available": available,
                "portfolio": portfolio,
            }
        except Exception as exc:
            logger.error("ecosystem_hermes_error", error=str(exc))
            return {"error": str(exc), "status": "unavailable"}

    @router.get("/organism", summary="Autonomous Organism status")
    async def organism_status() -> Dict[str, Any]:
        """Autonomous Organism service status."""
        try:
            from ai_multicolony.integrations.organism_bridge import OrganismBridge
            import os

            bridge = OrganismBridge(
                supabase_url=os.getenv("SUPABASE_URL", ""),
                supabase_anon_key=os.getenv("SUPABASE_ANON_KEY", ""),
            )
            available = await bridge.is_available()
            result: Dict[str, Any] = {"available": available}
            if available:
                # Try to get status for a default org
                org_id = os.getenv("DEFAULT_ORG_ID", "")
                if org_id:
                    status = await bridge.get_organism_status(org_id)
                    result["organism"] = status.model_dump()
            await bridge.close()
            return result
        except Exception as exc:
            logger.error("ecosystem_organism_error", error=str(exc))
            return {"error": str(exc), "status": "unavailable"}

    @router.get("/overview", summary="Combined ecosystem overview")
    async def ecosystem_overview() -> Dict[str, Any]:
        """Combined dashboard data from all ecosystem services.

        This is the primary endpoint for the unified dashboard,
        aggregating data from Crucix, HermesQuant, and Autonomous Organism.
        """
        overview: Dict[str, Any] = {
            "ecosystem": "AI-MultiColony-Ecosystem",
            "version": "3.0.0",
            "crucix": {},
            "hermes_quant": {},
            "autonomous_organism": {},
        }

        # Crucix data
        try:
            from ai_multicolony.integrations.crucix_client import CrucixClient

            client = CrucixClient()
            sweep = await client.get_sweep_data()
            health = await client.get_health()
            await client.close()
            overview["crucix"] = {
                "regime": sweep.regime,
                "sources_active": sweep.sources_active,
                "signal_count": len(sweep.signals),
                "news_count": len(sweep.news),
                "health": health.status,
            }
        except Exception:
            overview["crucix"] = {"status": "unavailable"}

        # HermesQuant data
        try:
            from ai_multicolony.integrations.hermes_bridge import HermesQuantBridge

            bridge = HermesQuantBridge()
            portfolio = await bridge.get_portfolio_status()
            overview["hermes_quant"] = portfolio
        except Exception:
            overview["hermes_quant"] = {"status": "unavailable"}

        # Organism data
        try:
            from ai_multicolony.integrations.organism_bridge import OrganismBridge
            import os

            bridge = OrganismBridge(
                supabase_url=os.getenv("SUPABASE_URL", ""),
                supabase_anon_key=os.getenv("SUPABASE_ANON_KEY", ""),
            )
            available = await bridge.is_available()
            overview["autonomous_organism"] = {"available": available}
            await bridge.close()
        except Exception:
            overview["autonomous_organism"] = {"status": "unavailable"}

        return overview

    return router

File: dashboard/src/lib/ecosystem.ts
/**
 * Ecosystem Client — Typed frontend client for all ecosystem services.
 *
 * Provides React hooks and API methods for:
 *   - Crucix OSINT intelligence
 *   - HermesQuant trading
 *   - Autonomous Organism
 *   - Ecosystem overview
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface EcosystemStatus {
  ecosystem: string;
  version: string;
  overall: "healthy" | "degraded";
  services: {
    crucix?: { status: string; type: string; port?: number };
    hermes_quant?: { status: string; type: string };
    autonomous_organism?: { status: string; type: string };
  };
}

export interface CrucixSweepData {
  regime: string;
  sources_active: number;
  signals: Array<{
    id: string;
    source: string;
    category: string;
    summary: string;
    delta?: { direction: string; magnitude: number };
  }>;
  news: Array<{ title: string; source: string; timestamp: string }>;
  market_indices: Record<string, number>;
  commodities: Record<string, number>;
}

export interface HermesPortfolio {
  available: boolean;
  portfolio?: {
    allocation?: unknown;
    journal_stats?: unknown;
    pnl?: unknown;
  };
  error?: string;
}

export interface OrganismInfo {
  available: boolean;
  organism?: {
    org_id: string;
    name: string;
    generation: number;
    status: string;
  };
}

export interface EcosystemOverview {
  ecosystem: string;
  version: string;
  crucix: {
    regime?: string;
    sources_active?: number;
    signal_count?: number;
    news_count?: number;
    health?: string;
    status?: string;
  };
  hermes_quant: Record<string, unknown>;
  autonomous_organism: { available?: boolean; status?: string };
}

// ---------------------------------------------------------------------------
// API Client
// ---------------------------------------------------------------------------

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

async function fetchAPI<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { "Content-Type": "application/json", ...init?.headers },
    ...init,
  });
  if (!res.ok) {
    throw new Error(`API error: ${res.status} ${res.statusText}`);
  }
  return res.json();
}

/** Fetch overall ecosystem status. */
export async function getEcosystemStatus(): Promise<EcosystemStatus> {
  return fetchAPI<EcosystemStatus>("/api/ecosystem/status");
}

/** Fetch Crucix OSINT sweep data. */
export async function getCrucixData(): Promise<{
  sweep: CrucixSweepData;
  health: { status: string; uptime: number };
}> {
  return fetchAPI("/api/ecosystem/crucix");
}

/** Fetch HermesQuant trading status. */
export async function getHermesStatus(): Promise<HermesPortfolio> {
  return fetchAPI<HermesPortfolio>("/api/ecosystem/hermes");
}

/** Fetch Autonomous Organism status. */
export async function getOrganismStatus(): Promise<OrganismInfo> {
  return fetchAPI<OrganismInfo>("/api/ecosystem/organism");
}

/** Fetch combined ecosystem overview. */
export async function getEcosystemOverview(): Promise<EcosystemOverview> {
  return fetchAPI<EcosystemOverview>("/api/ecosystem/overview");
}
